### PR TITLE
Add `visualization.matplotlib.plot_pareto_front`

### PR DESCRIPTION
**Pull request by tohmae**
This PR addresses issue #2113 

pareto_front
.

### CODE REVIEW

File: optuna/visualization/matplotlib/_pareto_front.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+
+if _imports.is_successful():
+    from optuna.visualization.matplotlib._matplotlib_imports import Axes
+    from optuna.visualization.matplotlib._matplotlib_imports import plt
+
+_logger = optuna.logging.get_logger(__name__)
+
+
+@experimental("2.4.0")
```
Comment (nzw0301) [inline optuna/visualization/matplotlib/_pareto_front.py:19]: I think it should be the next release version instead of `2.4.0`.

File: optuna/visualization/matplotlib/_pareto_front.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    target_names: Optional[List[str]],
+    include_dominated_trials: bool = False,
+    axis_order: Optional[List[int]] = None,
+) -> "Axes":
+
+    # Set up the graph style.
+    plt.style.use("ggplot")  # Use ggplot style sheet for similar outputs to plotly.
+    fig = plt.figure()
+    ax = fig.add_subplot(projection="3d")
+    #    _, ax = plt.subplots(projection='3d')
```
Comment (nzw0301) [inline optuna/visualization/matplotlib/_pareto_front.py:177]: Could you remove this line?

File: optuna/visualization/matplotlib/_pareto_front.py (hunk 1 above)
Comment (crcrpar) [inline optuna/visualization/matplotlib/_pareto_front.py:19]:
```suggestion
@experimental("2.6.0")
```

File: optuna/visualization/matplotlib/_pareto_front.py (hunk 2 above)
Comment (crcrpar) [inline optuna/visualization/matplotlib/_pareto_front.py:177]:
```suggestion
```

File: optuna/visualization/matplotlib/_pareto_front.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+                f"Given `axis_order` {axis_order} contains invalid index {max(axis_order)} "
+                "higher than 1."
+            )
+        if min(axis_order) < 0:
+            raise ValueError(
+                f"Given `axis_order` {axis_order} contains invalid index {min(axis_order)} "
+                "lower than 0."
+            )
+
+    ax.scatter(
```
Comment (nzw0301) [inline optuna/visualization/matplotlib/_pareto_front.py:148]: The plot shows `trial` label without any associated points when `include_dominated_trials=False`. (See the python code black to plot and generated figure.)
Plotly's plot does not show `trial` label with the same setting. I would suggest adding `if include_dominated_trials:`:  

--- 

```python
    if include_dominated_trials:
        ax.scatter(
            x=[t.values[axis_order[0]] for t in trials[len(study.best_trials) :]],
            y=[t.values[axis_order[1]] for t in trials[len(study.best_trials) :]],
            color=cmap(0),
            label="Trial",
        )
```

The code block above plots the following figure that contains `trial` label but no associated points.

```python
import optuna


def objective(trial):
    x = trial.suggest_float("x", 0, 5)
    y = trial.suggest_float("y", 0, 3)

    v0 = 4 * x ** 2 + 4 * y ** 2
    v1 = (x - 5) ** 2 + (y - 5) ** 2
    return v0, v1


study = optuna.create_study(directions=["minimize", "minimize"])
study.optimize(objective, n_trials=50)

optuna.visualization.matplotlib.plot_pareto_front(study, include_dominated_trials=False)
```
download (1)

File: optuna/visualization/matplotlib/_pareto_front.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+                f"Given `axis_order` {axis_order} contains invalid index {max(axis_order)} "
+                "higher than 2."
+            )
+        if min(axis_order) < 0:
+            raise ValueError(
+                f"Given `axis_order` {axis_order} contains invalid index {min(axis_order)} "
+                "lower than 0."
+            )
+
+    ax.scatter(
```
Comment (nzw0301) [inline optuna/visualization/matplotlib/_pareto_front.py:219]: Same as the 2D case of  the`trial` issue.

File: tests/visualization_tests/matplotlib_tests/test_pareto_front.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        )
+
+    target_names = ["Foo", "Bar", "Baz"]
+    figure = plot_pareto_front(study=study, target_names=target_names, axis_order=axis_order)
+
+    assert figure.has_data()
+
+
+@pytest.mark.parametrize("include_dominated_trials", [False, True])
+def test_plot_pareto_front_unsupported_dimensions(include_dominated_trials: bool) -> None:
```
Comment (nzw0301) [inline tests/visualization_tests/matplotlib_tests/test_pareto_front.py:142]: Since this test contains supported dimension cases: `n_objectives == 2` and `n_objectives == 3`, this test name is supposed to be inaccurate. I know the same test name exists in Plotly's test https://github.com/optuna/optuna/blob/62dc24574acd13d76ff1a4a94303a3302e5e3fbe/tests/visualization_tests/test_pareto_front.py#L212.

Personally, I suggest changing this test name to `test_plot_pareto_front_dimensions`, but I may miss some context of this name, so I would like to hear the opinions from the core-team members.

**Review by nzw0301 — CHANGES_REQUESTED** [review]
It looks great! I leave some minor comments, but you can ignore the comments if the core-team members disagree with my comments because I'm not an official reviewer :)

**Comment by toshihikoyanase** [comment]
@crcrpar @nzw0301 Thank you for your reviews. Let me assign you to this PR.

**Comment by tohmae** [comment]
Thank you for your comments. I fixed the parts you pointed out. Could you please review again?

File: optuna/visualization/matplotlib/_pareto_front.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        label="Trial",
+    )
+    ax.scatter(
+        x=[t.values[axis_order[0]] for t in trials[: len(study.best_trials)]],
+        y=[t.values[axis_order[1]] for t in trials[: len(study.best_trials)]],
+        color=cmap(3),
+        label="Best Trial",
+    )
+
+    if include_dominated_trials is True:
```
Comment (nzw0301) [inline optuna/visualization/matplotlib/_pareto_front.py:161]:
```suggestion
    if include_dominated_trials:
```

File: optuna/visualization/matplotlib/_pareto_front.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    )
+    ax.scatter(
+        xs=[t.values[axis_order[0]] for t in trials[: len(study.best_trials)]],
+        ys=[t.values[axis_order[1]] for t in trials[: len(study.best_trials)]],
+        zs=[t.values[axis_order[2]] for t in trials[: len(study.best_trials)]],
+        color=cmap(3),
+        label="Best Trial",
+    )
+
+    if include_dominated_trials is True:
```
Comment (nzw0301) [inline optuna/visualization/matplotlib/_pareto_front.py:234]:
```suggestion
    if include_dominated_trials:
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
Thank you for updating the codes @crcrpar and I mentioned. Could you update the two lines again?

**Comment by tohmae** [comment]
Thank you for your comments. I updated.

File: tests/visualization_tests/matplotlib_tests/test_pareto_front.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    # Supported: n_objectives == 3.
+    study = optuna.create_study(directions=["minimize", "minimize", "minimize"])
+    study.optimize(lambda t: [0, 0, 0], n_trials=1)
+    figure = plot_pareto_front(study=study, include_dominated_trials=include_dominated_trials)
+    assert figure.has_data()
+
+    # Unsupported: n_objectives == 4.
+    with pytest.raises(ValueError):
+        study = optuna.create_study(directions=["minimize", "minimize", "minimize", "minimize"])
+        plot_pareto_front(study=study, include_dominated_trials=include_dominated_trials)
```
Comment (nzw0301) [inline tests/visualization_tests/matplotlib_tests/test_pareto_front.py:169]:
```suggestion
        study.optimize(lambda t: [0, 0, 0, 0], n_trials=1)
        plot_pareto_front(study=study, include_dominated_trials=include_dominated_trials)
```

Comment (nzw0301) [inline tests/visualization_tests/matplotlib_tests/test_pareto_front.py:169]:
```suggestion
        study.optimize(lambda t: [0, 0, 0, 0], n_trials=1)
        plot_pareto_front(study=study, include_dominated_trials=include_dominated_trials)
```
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
Sorry, one more thing

**Comment by nzw0301** [comment]
I think `test_plot_pareto_front_2d` and `test_plot_pareto_front_3d`  don't have some test cases in  Plotly's tests. I think they should be included as well.

**Comment by nzw0301** [comment]
Hi, @tohmae how do you think about my comment? I think the missing test functions can be added after this pull request as a follow-up pull request if it is difficult to add them.

**Comment by tohmae** [comment]
Thank you for your comments. I would be grateful if you could merge this pull request once.

File: tests/visualization_tests/test_pareto_front.py [hunk 9]
```diff
@@ -209,7 +209,7 @@ def test_plot_pareto_front_3d(
 
 
 @pytest.mark.parametrize("include_dominated_trials", [False, True])
-def test_plot_pareto_front_unsupported_dimensions(include_dominated_trials: bool) -> None:
+def test_plot_pareto_front_dimensions(include_dominated_trials: bool) -> None:
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_pareto_front.py:212]: Thank you for renaming the test function, but I feel this change is out of scope of this PR.
I think we have design choices of this test case. For example, we can also keep the current name, and remove the assertion of supported dimensions.
So, could you revert this change? 
```suggestion
def test_plot_pareto_front_unsupported_dimensions(include_dominated_trials: bool) -> None:
```

File: optuna/visualization/matplotlib/_pareto_front.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+
+if _imports.is_successful():
+    from optuna.visualization.matplotlib._matplotlib_imports import Axes
+    from optuna.visualization.matplotlib._matplotlib_imports import plt
+
+_logger = optuna.logging.get_logger(__name__)
+
+
+@experimental("2.7.0")
```
Comment (toshihikoyanase) [inline optuna/visualization/matplotlib/_pareto_front.py:19]: Could you update this version number?
```suggestion
@experimental("2.8.0")
```

**Comment by nzw0301** [comment]
@tohmae Thank you for accepting my suggestion! Let us merge this pull request after we apply @toshihikoyanase's suggestions to this pull request.

File: optuna/visualization/matplotlib/_pareto_front.py (hunk 10 above)
Comment (nzw0301) [reply optuna/visualization/matplotlib/_pareto_front.py:19]: Nice catch! Thanks!

File: tests/visualization_tests/test_pareto_front.py (hunk 9 above)
Comment (nzw0301) [reply tests/visualization_tests/test_pareto_front.py:212]: @toshihikoyanase I agree with you. It will be another follow-up pull request.

@tohmae  Sorry for confusing you. I should have clarified my suggestion.

**Comment by nzw0301** [comment]
Let me summarise the content of the follow-up pull requests

## plotly test
- add missing test functions: `test_plot_pareto_front_2d` and `test_plot_pareto_front_3d`
- Extract supported dimenstion cases from `test_plot_pareto_front_unsupported_dimensions`, and create another test for the extracted cases

## matplotlib test
- extract supported dimenstion cases from `test_plot_pareto_front_unsupported_dimensions`, and create another test for the extracted cases

**Review by nzw0301 — APPROVED** [review]

File: optuna/visualization/matplotlib/_pareto_front.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    ax.set_title("Pareto-front Plot")
+    cmap = plt.get_cmap("tab10")  # Use tab10 colormap for similar outputs to plotly.
+
+    if target_names is None:
+        target_names = ["Objective 0", "Objective 1"]
+    elif len(target_names) != 2:
+        raise ValueError("The length of `target_names` is supposed to be 2.")
+
+    ax.set_xlabel(target_names[0])
+    ax.set_ylabel(target_names[1])
```
Comment (toshihikoyanase) [inline optuna/visualization/matplotlib/_pareto_front.py:116]: When I changed the axis order, the labels are not shown as expected. I think we need to refer to the values of the `axis_order` list similarly to the [plotly-based implementation](https://github.com/optuna/optuna/blob/3b8a5fe71dcd511d73070a43394a9e70341b4db9/optuna/visualization/_pareto_front.py#L173-L174).

```python
ax.set_xlabel(target_names[axis_order[0]]
ax.set_ylabel(target_names[axis_order[1]]
```

`axis_order=[0, 1]` (default)
image

`axis_order=[1, 0]`: The axis labels are not swapped.
image

File: optuna/visualization/matplotlib/_pareto_front.py [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+    cmap = plt.get_cmap("tab10")  # Use tab10 colormap for similar outputs to plotly.
+
+    if target_names is None:
+        target_names = ["Objective 0", "Objective 1", "Objective 2"]
+    elif len(target_names) != 3:
+        raise ValueError("The length of `target_names` is supposed to be 3.")
+
+    ax.set_xlabel(target_names[0])
+    ax.set_ylabel(target_names[1])
+    ax.set_zlabel(target_names[2])
```
Comment (toshihikoyanase) [inline optuna/visualization/matplotlib/_pareto_front.py:188]: ditto.

**Review by toshihikoyanase — APPROVED** [review]
I found small bugs about the `axis_order` arguments. They should be fixed in a follow-up PR.
Overall, I agree with @nzw0301, and it looks good to me.
